src/interrupt.rs

```diff
@@ -663,6 +663,28 @@ pub mod asynch {
     }
 }
 
+#[cfg(feature = "critical-section-isr")]
+pub mod critical_section {
+
+    static CS_GUARD: std::sync::Mutex<Option<IsrCriticalSectionGuard>> = Mutex::new(None);
```

Comment:
We cannot use a rust std Mutex on a ISR related CriticalSection. You have to keep in mind that interrupts execution can happen recursively, and its not allowed to block in a ISR itself, as it would crash the system in a short amount of time, and a std Mutex have no mechanism to overcome this situation.

Reply:
I suggest making `enter` and `exit` public here https://github.com/esp-rs/esp-idf-hal/blob/e30d7ab708065d48f8a96e32993bde3ff3946370/src/interrupt.rs#L172-L215

That way, we can use those instead of the Mutex, which would also mirror https://github.com/esp-rs/esp-idf-hal/blob/e30d7ab708065d48f8a96e32993bde3ff3946370/src/task.rs#L775

Reply:
I've given this a try, I need to test it, and maybe make an example that uses it.

Reply:
> Maybe enter and exit should be public, so that people can use them to implement their own CS implementation?

In general the hal/svc crate's aim to provide a usable API the enduser directly can use. Providing something in the public API that can be easily misused is not a good thing. 

If a user wants to implement its own CS implementation he should so directly ontop of esp-idf-sys.

That said its totally ok and fine if we provide good critical-section implementation for different needs and i am fine with that. 

Though since differences in implementation are often too opaque for a general API user, any new addition should be clearly documented. Our past track-record is not good here, that's why i would appreciate if you could clearly state the intention of the API ( e.g what solution it offers as a doc-string) and in what situation it should/shouldn't used.

Reply:
Also keep in mind that the feature gated critical-section implementation is written for consumption with the critical-section crate. The critical-section crate goal in general is to provide a portable implementation, so good when you want to write a library that runs everywhere. But its goal is not to provide "the always best critical-section in every situation". 

Since ISR's are still extremely platform depended, its less logical to provide a critical-section crate implementation that makes only sense to use inside a non portable ISR, since the part of the code probably is in a binary specific path. So what i am saying is -> 
1) We should provide specialized critical-section implementation for the specific usage patterns that are directly depended on esp-idf features in some form. 
2) The critical-section crate itself doesn't give any specific guarantees so providing multiple implementations is may not the best path forward.

Reply:
> Though since differences in implementation are often too opaque for a general API user, any new addition should be clearly documented. Our past track-record is not good here, that's why i would appreciate if you could clearly state the intention of the API ( e.g what solution it offers as a doc-string) and in what situation it should/shouldn't used.

The intention of this critical section implementation is to allow `critical_section`  to be used during an interrupt context, disabling interrupts during time-critical code. The current one provided in `task.rs` is thread-local, and not safe for interrupt, as I understand (correct me if I am wrong on this?). 

An example of its use is seen here: 

https://github.com/DaneSlattery/hx711/blob/72cadcfeec373f8ec04c1e39b559cee3017c00dd/src/hx711.rs#L142C8-L151C12

In the above code, `read_bits` is called during an interrupt, as in this example:
https://github.com/DaneSlattery/hx711/blob/72cadcfeec373f8ec04c1e39b559cee3017c00dd/examples/interrupt.rs#L81C5-L88C8



The ESP-idf-hal is already split on two implementations for critical sections. The default one in `task.rs` and another in `interrupt.rs`. However, only the one in `task.rs` is considered for the default implementation exposed in this library, but it is not always suitable.

Reply:
@DaneSlattery @beeb 

The problem with the `critical-section` crate is that it has a too simplistic view of the world, where it thinks a Rust MCU SDK would only have one type of "critical section" (a.k.a. blocking mutex).

While for bare-metal SDKs which don't really have an RTOS (like most Embassy HALs) that might be the case, this is just not true for RTOS SDKs like the `esp-idf-*` crates.

These crates have at least three candidates for "a critical section"
- A regular RTOS recursive mutex (roughly the same as `std::sync::Mutex` except recursive). This is what `esp-idf-hal` exposes as a critical section impl to the `critical-section` crate. Additionally, this critical section is exposed with the `esp-idf-hal`-specific API `CriticalSection`. But, as you have noticed, it cannot prevent interrupts from happening. Note also that your terminology is incorrect. This critical section is definitely not "thread local" as thread-local means something else (look at `NoopRawMutex` in `embassy-sync` for details). It is just a regular RTOS mutex, which happens to be - in 99% of the cases - what users would probably need, hence why it is exposed as THE `critical-section` impl in `esp-idf-hal`.
- A "stop the RTOS scheduler" type of critical section. This critical section would not stop interrupts from happening, but would suspend task preemption. `esp-idf-hal` does not expose this guy.
- A "stop all interrupts" type of critical section. This critical section is available in `esp-idf-hal` under the name `IsrCriticalSection` but is NOT exposed as an alternative implementation for the `critical-section` crate. Not also that on multi-core MCUs like the xtensa-based ESP32 and ESP32S3 this critical section only stops the interrupts on _one_ of the cores, while the other might still get interrupts and will do spin-locking if it tries to get access to the critical section, but that's a detail

OK, where am I getting with all that? 
I'm not sure that `esp-idf-hal` should provide an alternative `critical-section` implementation of type (3) from above. 

There is nothing in the `critical-section` crate which says that the critical section impl should disable all interrupts. Quite the opposite. The `critical-section` crate foes at length to explain in its docu, that the exposed critical section for STD environments might just be a regular STD mutex, or suchlike.

I would actually point the finger at the `hx711` crate. :) (Sorry for that!!) It blindly assumes that what `critical-section` exposes as a critical section will stop all interrupts, which is just not the case!

I would say, if the `hx711` crates wants to broaden its compatibility, it should switch to using `embassy-sync::blocking_mutex::RawMutex` instead of  (or in addition to) its existing usage of `critical-section`. Thus, user would be free to provide a `RawMutex` implementation which indeed has the capabilities this crate needs (as in "stop all interrupts"), as the `RawMutex` trait - unlike `critical-section` - might have multiple implementations.

The other question I have (and here I'm really not sure) is: are we really sure that we DO need "disable all interrupts" type of behavior for `hx711`? I do realize that the crate does bit-banging, but have we tried raising the priority of the thread that does the bit-banging stuff to as-high as possible, and then trying out? Maybe that would be good enough and that would mean folks would be OK with the existing `hx711` impl?

Reply:
Also - and if that's not clear - switching to `RawMutex` will not mean users of baremetal crates would have a harder time using the `hx711` crate, as `embassy-sync` has an out of the box implementation of `RawMutex` called `CriticalSectionRawMutex` which is based on... the mutex provided by the `critical-section` crate.

And to complete the picture, `esp-idf-hal`s  `IsrCriticalSection` (the one you do need for your crate and the one you do try to expose as an alternative implementation for the `critical-section` crare) already does implement the `RawMutex` contract from `embassy-sync`.

Reply:
@ivmarkov What a well written response, thank you.

When I first wrote the `hx711` crate, it did assume that `critical-section` impl would stop all interrupts, because that is what it does in the bare-metal `esp-hal`. The trouble comes in with `std` targets such as the `esp-idf-hal`, where the scheduler may still run tasks, or even run interrupts with a higher priority. This is a problem for the time-sensitive bit-banging. Switching to `embassy-sync::blocking_mutex::RawMutex` for this is certainly a good idea, given it holds more restrictive assumptions in both `esp-idf-hal` and `esp-hal` impls.

I do think that raising the priority of the bit-banging task could also yield positive results in the RTOS usages of the `hx711` crate. I have not played around with this too much, but I assume even without an executor, 
```
ThreadSpawnConfiguration {
        name: Some("hi-lvl-thread".as_bytes()),
        stack_size: 1000,
        priority: 15,
        ..Default::default()
    }
    .set()
    .unwrap();
``` 
would work. Do you know what the priority of `main()` is? 

I think given the recommendations, we can close this PR. @beeb , what do you think?

Reply:
> Do you know what the priority of main() is?

Almost the lowest possible and no possibility to raise it. :)

Reply:
@DaneSlattery I agree that some nice alternatives have been provided here and we can probably close this PR :) Thanks @Vollbrecht and @ivmarkov for the helpful insights!